scrapy/contrib/downloadermiddleware/redirect_offsite.py

```diff
@@ -0,0 +1,34 @@
+import re
+from scrapy import log
+from scrapy.http import Request
+from scrapy.exceptions import IgnoreRequest
+from scrapy.utils.httpobj import urlparse_cached
+from scrapy.contrib.downloadermiddleware.redirect import RedirectMiddleware
+
+
+class RedirectOffsiteMiddleware(RedirectMiddleware):
+
+    def process_response(self, request, response, spider):
+        redirected = super(RedirectOffsiteMiddleware, self).process_response(request, response, spider)
+        return self.offsite_filtered(redirected, spider)
+
+    def offsite_filtered(self, redirected, spider):
+
+        allowed_domains = getattr(spider, 'allowed_domains', None)
+
+        if not allowed_domains:
+            return redirected
+
+        if isinstance(redirected, Request):
```

Comment:
From the documentation: 
```process_response()``` should either: return a ```Response``` object, return a ```Request``` object or raise a ```IgnoreRequest``` exception.
(http://doc.scrapy.org/en/latest/topics/downloader-middleware.html#scrapy.contrib.downloadermiddleware.DownloaderMiddleware.process_response)
So there is no problem to return a ```Request``` from ```process_response```.
In ```RedirectMiddleware.process_response``` method htat is used here there is a number of cases when the response contains a redirection, so a new ```Request``` is returned that is pointing to the new destination.